### DISCUSSION TITLE
Question about Sampling Multiple Edge Types Simultaneously in Heterogeneous Graphs using PyG

### DISCUSSION THREAD

**Discussion by ccoverflow** [pyg-team/pytorch_geometric#7551] [Q&A]
Hi, I have a heterogeneous graph data that has been converted to the HeteroData class. This heterogeneous graph data has more than 25 types of edges, each with a timestamp. I need to sample multiple types of edges simultaneously when loading (for example, for a type 1 edge, select other edges within its designated hops (like 2 hops), which have a common parent node and whose occurrence time satisfies the chronological order, to be the neighbors of edge 1, and set to sample n neighbors of edge 1). However, the 'edge_label_index' parameter in 'torch_geometric.loader.link_neighbor_loader' can only accept one type of edge. I have also tried implemented loaders such as HGTLoader, but none of them can achieve the effect I want. I'm wondering if there are similar class methods in PyG that can achieve this goal?

<img width="388" alt="Pasted Graphic" src="https://github.com/pyg-team/pytorch_geometric/assets/45088469/c69530be-9c26-4fdb-90bb-6972d4b77000">

**Comment by wsad1** [comment]
Currently `LinkNeighborLoader` can only start sampling from on edge type. And it doesn't support edges other than the seed edge having a timestamp.
But there is an open PR to implement sampling from multiple node types https://github.com/pyg-team/pytorch_geometric/pull/5521, cc: @Padarn.
And support for edge time in loaders is being worked on.

**Reply by ccoverflow** [reply]
“Currently `LinkNeighborLoader` can only start sampling from on edge type.“Do you mean that it can only sample the same type of edge at a time

**Reply by rusty1s** [reply]
Yeah, I think @wsad1 meant "from one edge type". This is a current limitation, but it's non-trivial to support multiple edge types here, especially because the decoder now needs to operate on partially available edge types. I think if you really wanna decode multiple edge types, it makes more sense to convert your graph into a typed homogeneous one first via `data.to_homogeneous()`.

**Reply by ccoverflow** [reply]
Thank you for your answer. Do you mean that after homogeneous graph conversion, edge_type can also be used to implement different types of edge sampling, which will be more efficient?

**Reply by rusty1s** [reply]
Yeah, in that case you are basically free to start sampling from any edge in your graph (which may be of different type as well, i.e., `edge_type` does not need to point to a single index only).

**Reply by ccoverflow** [reply]
Thank you very much for your advice. I'll give it a try.

**Reply by ccoverflow** [reply]
@rusty1s  My "HeteroData" has a label property on the edge of each type, where label values are of two types, 0 and 1. After I use "data.to_homogeneous()" how do I add the label properties of each type of edge into converted labels of the same composition?
<img width="980" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/45088469/5f9d028d-6a42-4b13-8de4-ea43542e38f7">

**Reply by rusty1s** [reply]
Sorry for late reply. It's hard for me to understand your issue. Do you have a small example to illustrate? In the end `label` should just be concatenated across edge types when calling `to_homogeneous()`.
